crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -182,211 +176,392 @@ const MAX_RECURSION_DEPTH: usize = 24;
-    constraints: IndexVec<ScopedVisibilityConstraintId, VisibilityConstraint<'db>>,
-}
-
-impl Default for VisibilityConstraintsBuilder<'_> {
-    fn default() -> Self {
-        Self {
-            constraints: IndexVec::from_iter([
-                VisibilityConstraint(VisibilityConstraintInner::AlwaysTrue),
-                VisibilityConstraint(VisibilityConstraintInner::AlwaysFalse),
-                VisibilityConstraint(VisibilityConstraintInner::Ambiguous),
-            ]),
-        }
-    }
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
+    constraint_cache: FxHashMap<Constraint<'db>, u32>,
+    interior_cache: FxHashMap<InteriorNode, u32>,
+    not_cache: FxHashMap<ScopedVisibilityConstraintId, ScopedVisibilityConstraintId>,
+    and_cache: FxHashMap<
+        (ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+        ScopedVisibilityConstraintId,
+    >,
+    or_cache: FxHashMap<
+        (ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+        ScopedVisibilityConstraintId,
+    >,
 }
 
 impl<'db> VisibilityConstraintsBuilder<'db> {
     pub(crate) fn build(self) -> VisibilityConstraints<'db> {
         VisibilityConstraints {
             constraints: self.constraints,
+            interiors: self.interiors,
         }
     }
 
-    fn add(&mut self, constraint: VisibilityConstraintInner<'db>) -> ScopedVisibilityConstraintId {
-        self.constraints.push(VisibilityConstraint(constraint))
+    fn interior_node(&self, a: ScopedVisibilityConstraintId) -> InteriorNode {
+        debug_assert!(!a.is_terminal());
+        self.interiors[a.0 as usize]
+    }
+
+    /// Returns whether `a` or `b` has a "larger" atom. TDDs are ordered such that interior nodes
+    /// can only have edges to "larger" nodes. Terminals are considered to have a larger atom than
+    /// any internal node, since they are leaf nodes.
+    fn cmp_atoms(
+        &self,
+        a: ScopedVisibilityConstraintId,
+        b: ScopedVisibilityConstraintId,
+    ) -> Ordering {
+        if a == b {
+            Ordering::Equal
+        } else if a.is_terminal() {
+            Ordering::Greater
+        } else if b.is_terminal() {
+            Ordering::Less
+        } else {
+            self.interior_node(a).atom.cmp(&self.interior_node(b).atom)
+        }
+    }
+
+    /// Adds a constraint, ensuring that we only store any particular constraint once.
+    #[allow(clippy::cast_possible_truncation)]
+    fn add_constraint(&mut self, constraint: Constraint<'db>, copy: u8) -> Atom {
+        let index = *self.constraint_cache.entry(constraint).or_insert_with(|| {
+            let index = self.constraints.len() as u32;
+            self.constraints.push(constraint);
+            index
+        });
+        Atom::from_index_and_copy(index, copy)
+    }
+
+    /// Adds an interior node, ensuring that we always use the same visibility constraint ID for
+    /// equal nodes.
+    #[allow(clippy::cast_possible_truncation)]
```

Comment:
Is it possible to silence this on just the one line, instead of the whole function?

Reply:
These went away with the switch to `IndexVec`